Fix translation in bitwise operators and check if null before tolower call

oscarlocatelli opened the pull request:
Hi Garvin,

I think I have fixed a problem in GenerateEntityFilter that appears with MySql (I'm using Pomelo provider).

When I use a nullable column with some NULL values the translated LOWER(column) fail and the row is skipped by MySql.

I've added a null check in AND to skip the LOWER call.

```csharp
// Check if it is nullable
var defValue = prop.PropertyType.GetTypeInfo().IsValueType ? Activator.CreateInstance(prop.PropertyType) : null;
if (defValue == null)
{
    var isNotNull = Expression.NotEqual(Expression.Property(paramExpression, prop), Expression.Constant(null));
    globalTest = Expression.And(isNotNull, globalTest);
}
```

So from this:

```sql
| (LOCATE(CONVERT('search' USING utf8mb4) COLLATE utf8mb4_bin, LOWER(`u`.`user`)) > 0))
```

now it traslates in:

```sql
 | (`u`.`user` IS NOT NULL & (LOCATE(CONVERT('search' USING utf8mb4) COLLATE utf8mb4_bin, LOWER(`u`.`user`)) > 0)))
```

But this fail because of the **BITWISE AND** from Expression.And.
So I changed my Expression.And into Expression.AndAlso and your Expression.Or to Expression.OrElse, too.

Final translation is:

```sql
 OR (`u`.`user` IS NOT NULL AND (LOCATE(CONVERT('search' USING utf8mb4) COLLATE utf8mb4_bin, LOWER(`u`.`user`)) > 0)))
```

I hope I haven't created bugs and you could merge it.

garvincasimir commented:
Hey @oscarlocatelli  thank you very much for contributing. Is it possible for you to add a test in MysqlEntityTests.cs to validate this issue? If not, I can do so later. Sorry for ci failing, I was trying to upgrade the tests to entity framework 3 (should have done that in a different branch). I will fix that. As long as all tests pass we can approve.

garvincasimir commented:
@oscarlocatelli are you saying that there is a match in one of the columns but because another column is null, that row is skipped?

oscarlocatelli commented:
@garvincasimir Creating test I found that the only problem was the Bitwise OR..

With Bitwise OR if one of the searched column was null the corresponding row skipped.
With `Expression.OrElse` it seems to work correctly.

I reverted not significant edit and I added the test.

To complete the test I had to install Pomelo 3.1.1 and eliminate the InternalServiceProvider.
But I didn't upload this.

If it can be useful:


```xml
<PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="3.1.1" />
```

```csharp
//var serviceProvider = new ServiceCollection()
//     .AddEntityFrameworkMySql()
//     .BuildServiceProvider();

 var builder = new DbContextOptionsBuilder<PersonContext>();
     builder.UseMySql(@"server=mysql;database=dotnettest;user=tester;password=Rea11ytrong_3")
      //.UseInternalServiceProvider(serviceProvider)
      .UseLoggerFactory(consoleLogger);
```

oscarlocatelli commented:
Thanks for the merge.
